File: typescript/lld/cache-design/rate-limiter/RateLimiter.ts
import { ClientState } from './ClientState';
import { RateLimitResult, RateLimitStrategy } from './RateLimitStrategy';

// Context — the only class callers touch. Holds one strategy + per-client
// state so unrelated clients are never serialized against each other.
export class RateLimiter {
    private clientStates = new Map<string, ClientState>();

    constructor(private strategy: RateLimitStrategy) {}

    isAllowed(clientId: string, now: number = Date.now()): RateLimitResult {
        let state = this.clientStates.get(clientId);
        if (!state) {
            state = this.strategy.initialState();               // starts FULL, not empty
            this.clientStates.set(clientId, state);
        }
        // real code: acquire a per-client lock here before mutating state
        return this.strategy.tryConsume(state, now);
    }
}

File: python/lld/cache-design/rate-limiter/rate_limiter.py
from rate_limit_strategy import RateLimitStrategy


class RateLimiter:
    """Context — the only class callers touch. Holds one strategy + a
    per-client state map so unrelated clients are never serialized
    against each other."""

    def __init__(self, strategy: RateLimitStrategy):
        self._strategy = strategy
        self._client_states = {}

    def is_allowed(self, client_id: str, now: float):
        if client_id not in self._client_states:
            self._client_states[client_id] = self._strategy.initial_state()  # starts FULL
        state = self._client_states[client_id]
        # real code: acquire a per-client lock here before mutating state
        return self._strategy.try_consume(state, now)